### ISSUE TITLE
First character from file path gets truncated

### ISSUE THREAD

**Issue by CaBeckmann** [vanhauser-thc/AFLplusplus#205]
Hi, I tried to run multiple fuzzing instances in parallel over night and came back to all of them having crashed.

- Instance 1:
Started with: 
`$ AFL_PRELOAD=libbar.so AFL_TMPDIR=ramdisk/ LD_LIBRARY_PATH=. /root/AFLplusplus/afl-fuzz -i seeds/bin/ -o sync -e fuzzer1 -M fuzzer1 -p exploit -L 10 -R ./foo @@`
Crashed after: 3 hrs, 37 min, 43 sec
Error:
```
[-]  SYSTEM ERROR : Unable to create 'amdisk//.cur_input.fuzzer1'                                                      
    Stop location : write_to_testcase(), src/afl-fuzz-run.c:306                                                        
       OS message : No such file or directory
```
- Instance 2:
Started with:
`$ AFL_PRELOAD=libbar.so AFL_TMPDIR=ramdisk/ LD_LIBRARY_PATH=. /root/AFLplusplus/afl-fuzz -i seeds/bin/ -o sync/ -e fuzzer2 -S fuzzer2 -p explore -L 10 -R ./foo @@`
Crashed after: 1 hrs, 37 min, 42 sec
Error:
```
[-]  SYSTEM ERROR : Unable to open 'ync//fuzzer2/queue/id:000368,sync:fuzzer3,src:000320'
    Stop location : common_fuzzing(), src/afl-fuzz-one.c:4120
       OS message : No such file or directory
```
- Instance 3:
Started with:
`$ AFL_PRELOAD=libsbar.so AFL_TMPDIR=ramdisk/ LD_LIBRARY_PATH=. /root/AFLplusplus/afl-fuzz -i seeds/bin/ -o sync/ -e fuzzer3 -S fuzzer3 -p coe -L 10 -R ./foo @@`
Crashed after: 2 hrs, 17 min, 8 sec
Error:
```
Segmentation fault (core dumped)
```
- Instance 4:
Started with:
`$ AFL_PRELOAD=libbar.so AFL_TMPDIR=ramdisk/ LD_LIBRARY_PATH=. /root/AFLplusplus/afl-fuzz -i seeds/bin/ -o sync/ -e fuzzer4 -S fuzzer4 -p fast -L 10 -R ./foo @@`
Crashed after: 0 hrs, 3 min, 46 sec
Error:
```
afl-fuzz(56833) in realloc(): use after free 0x238c4d0c500
Abort trap (core dumped)
```

I tried to examine the core file for the segmentation fault. The core file for the "use after free" crash sadly did get overwritten by this one.
```
#0  0x00000927e1427dde in common_fuzzing (argv=0x7f7ffffe6a68, MOpt_globals=
      {finds = 0x927e14374f0, finds_v2 = 0x927e1437570, cycles = 0x927e1437370, cycles_v2 = 0x927e14373f0, cycles_v3 = 0x927e1437470, is_pilot_mode = 0, pTime = 0x927e1438cb0, period = 500000, havoc_stagename = 0x927e140334b "MOpt-core-ha
voc", splice_stageformat = 0x927e140278f "MOpt-core-splice %u", havoc_stagenameshort = 0x927e13fff09 "MOpt_core_havoc", splice_stagenameshort = 0x927e14052e1 "MOpt_core_splice"}) at src/afl-fuzz-one.c:3719
3719                  FLIP_BIT(out_buf, temp_len_puppet + 1);
(gdb) bt
#0  0x00000927e1427dde in common_fuzzing (argv=0x7f7ffffe6a68, MOpt_globals=
      {finds = 0x927e14374f0, finds_v2 = 0x927e1437570, cycles = 0x927e1437370, cycles_v2 = 0x927e14373f0, cycles_v3 = 0x927e1437470, is_pilot_mode = 0, pTime = 0x927e1438cb0, period = 500000, havoc_stagename = 0x927e140334b "MOpt-core-havoc", splice_stageformat = 0x927e140278f "MOpt-core-splice %u", havoc_stagenameshort = 0x927e13fff09 "MOpt_core_havoc", splice_stagenameshort = 0x927e14052e1 "MOpt_core_splice"}) at src/afl-fuzz-one.c:3719
#1  0x00000927e142d604 in fuzz_one (argv=Variable "argv" is not available.
) at alloc-inl.h:85
#2  0x00000927e1412995 in main (argc=16, argv=0x7f7ffffe69f8, envp=Variable "envp" is not available.
) at src/afl-fuzz.c:978
Current language:  auto; currently minimal
(gdb) p out_buf
$1 = (u8 *) 0x92a8aee4ff0 "\003e\034=\003\\" <Address 0x92a8aee5000 out of bounds>
(gdb) p len
$2 = 260
```
It looks like out_buf only has a length of 16, while len says it should have a length of 260.

I tried reproducing the behaviour by running the same instances again. After 3 hours all of the slaves had crashed again with the following error:
```
[-]  SYSTEM ERROR : Unable to open 'ync//fuzzer2/queue/id:000529,sync:fuzzer3,src:000536'
    Stop location : common_fuzzing(), src/afl-fuzz-one.c:4120
       OS message : No such file or directory
```
The exact file name was different between the instances, but each was missing the first character. Also I noticed that each crashing file was synced from another instance. Of course with the high amount of synced files this may just be a coincidence.

Since all of the crashes happened during a MOpt stage I also ran each instance again without the -L parameter. They are currently all still running after 20 hours.

I am using the master branch with the latest commit being `cdbb35d` on OpenBSD 6.5.

If there is anything else I can do to help you debug this please let me know.

**Comment by andreafioraldi** [comment]
Hi, thank you for the report!
Can you rerun Instance 4 with MOpt and afl-fuzz built with AddressSanitizer? (build afl++ using `make ASAN_BUILD=1`).

**Comment by andreafioraldi** [comment]
@vanhauser-thc `-L 10 -R` is inconsistent, radamsa is not compatible with MOpt. We have to add a check that prevents the fuzzer to start if a not standard mutator is specified alongside MOpt.
The issue smells like an heap overflow, maybe in the MOpt code that, IIRC, you ported directly from the academic project, so there is an high probability that is broken for cases (like synced fuzzing) not evaluated in the paper.

**Comment by andreafioraldi** [comment]
@CaBeckmann we have to re-engineer the MOpt code cause at the moment is just the code ported from the original project and it is not integrated well into AFL++. Try to see if removing -R the bug is still present (use always ASAN_BUILD to have it triggered in a decent time, not after 3h). Thanks.

**Comment by CaBeckmann** [comment]
@andreafioraldi Sadly ASAN is not supported on OpenBSD. I am going the restart all instances with -L and without -R.

**Comment by andreafioraldi** [comment]
Oh sad, maybe try valgrind, I spotted issues in afl-fuzz with it many times.

**Comment by CaBeckmann** [comment]
Checking back after almost 3 hours: Instance 2 and 3 crashed again.

Instance 3:
```
 [-]  SYSTEM ERROR : Unable to open 'sync//fuzzer3/queue/id:000597,sync:fuzzer2,src:000597'                                                                                                                                                    
    Stop location : common_fuzzing(), src/afl-fuzz-one.c:4120
       OS message : File name too long
```
Same place but with a different error. The file path looks correct and is definitely not too long, so I am not sure what exactly went wrong.

Instance 2 segfaulted. Info from the core file:
```
#0  0x0000087807ecefda in common_fuzzing (argv=0x7f7ffffc6270, MOpt_globals=
      {finds = 0x87807ede4f0, finds_v2 = 0x87807ede570, cycles = 0x87807ede370, cycles_v2 = 0x87807ede3f0, cycles_v3 = 0x87807ede470, is_pilot_mode = 0, pTime = 0x87807edfcb0, period = 500000, havoc_stagename = 0x87807eaa34b "MOpt-core-ha
voc", splice_stageformat = 0x87807ea978f "MOpt-core-splice %u", havoc_stagenameshort = 0x87807ea6f09 "MOpt_core_havoc", splice_stagenameshort = 0x87807eac2e1 "MOpt_core_splice"}) at src/afl-fuzz-one.c:3727
3727                  FLIP_BIT(out_buf, temp_len_puppet + 1);
(gdb) p out_buf
$1 = (u8 *) 0x87ab669c1c0 "\020X#@P`p=\020 0@P`p\2005 \""
Current language:  auto; currently minimal
(gdb) p len
$2 = 514
(gdb) p temp_len_puppet 
$3 = 29183
```

Valgrind is also not supported on OpenBSD. There was a port with minimal functionality but it looks like it has been abandoned almost 4 years ago. I did not manage to get it to work on my machine.

**Comment by andreafioraldi** [comment]
```
[ENAMETOOLONG]
    A component of a pathname exceeded NAME_MAX characters, or an entire pathname (including the terminating NUL) exceeded PATH_MAX bytes.```

Try to see the value of those vars.

Try also to uncomment this line `// #define SIMPLE_FILES` in include/config.h, this should generate short names.

**Comment by andreafioraldi** [comment]
Regards the second crash, should be an issue of the MOpt code.
```c
            case 2:
              if (temp_len < 2) break;
              temp_len_puppet = UR(temp_len << 3);
              FLIP_BIT(out_buf, temp_len_puppet);
              FLIP_BIT(out_buf, temp_len_puppet + 1);
              FLIP_BIT(out_buf, temp_len_puppet + 2);
              FLIP_BIT(out_buf, temp_len_puppet + 3);
              MOpt_globals.cycles_v2[STAGE_FLIP4] += 1;
              break;
```

If temp_len_puppet is the last bit for instance, FLIP_BIT(out_buf, temp_len_puppet + 1); is out of bound.

In this case, you have to print temp_len (not len) and multiply it by 8. Then you have to check that `temp_len_puppet < (temp_len*8)`.

**Comment by CaBeckmann** [comment]
`#define NAME_MAX                  255`
`#define PATH_MAX                 1024`
Both of these are big enough that the path should not hit either.

Instance 1 and 4 have now also crashed. 
Instance 4 had the missing character in the path:
```
[-]  SYSTEM ERROR : Unable to open 'ync//fuzzer4/queue/id:000649,sync:fuzzer1,src:000219,+cov'
    Stop location : common_fuzzing(), src/afl-fuzz-one.c:4120
       OS message : No such file or directory
```
Instance 1 crashed because of "use after free":
```
afl-fuzz(62859) in free(): use after free 0xf74f70e8e00
Abort trap (core dumped)
```
This time I have the core file:
```
(gdb) bt
#0  thrkill () at -:3
#1  0x00000f74b1f11b9e in _libc_abort () at /usr/src/lib/libc/stdlib/abort.c:51
#2  0x00000f74b1f4c3b9 in wrterror (d=Variable "d" is not available.
) at /usr/src/lib/libc/stdlib/malloc.c:297
#3  0x00000f74b1f4ce37 in ofree (argpool=Variable "argpool" is not available.
) at tcb.h:43
#4  0x00000f74b1f4c4b2 in free (ptr=0xf74f70e8d00) at /usr/src/lib/libc/stdlib/malloc.c:1451
#5  0x00000f721f3728d7 in common_fuzzing (argv=0x7f7ffffbcd00, MOpt_globals=
      {finds = 0xf721f3828a0, finds_v2 = 0xf721f382b20, cycles = 0xf721f382020, cycles_v2 = 0xf721f3822a0, cycles_v3 = 0xf721f382520, is_pilot_mode = 1, pTime = 0xf721f382cc0, period = 50000, havoc_stagename = 0xf721f34923c "MOpt-havoc", 
splice_stageformat = 0xf721f349efa "MOpt-splice %u", havoc_stagenameshort = 0xf721f34e778 "MOpt_havoc", splice_stagenameshort = 0xf721f34b3d0 "MOpt_splice"}) at alloc-inl.h:108
#6  0x00000f721f377604 in fuzz_one (argv=Variable "argv" is not available.
) at alloc-inl.h:85
#7  0x00000f721f35c995 in main (argc=15, argv=0x7f7ffffbcc98, envp=Variable "envp" is not available.
) at src/afl-fuzz.c:978
```

I am now compiling afl++ with `SIMPLE_FILES` defined and your fix included.

**Comment by vanhauser-thc** [comment]
I am running afl-fuzz with MOpt -L0 and ASAN_BUILD in a few -M/-S setup. after an hour there is still no crash but I keep it running.

@CaBeckmann - do you still encounter crashes after using the fix from andrea?

**Comment by CaBeckmann** [comment]
It seems to be stable now. After 40 hours all 4 instances are still running.

**Comment by vanhauser-thc** [comment]
please repopen the issue if you encounter any other issue :-)

**Comment by vanhauser-thc** [comment]
Ooops reopening it again ... because there could be an additional problem with filenames on openbsd.

@CaBeckmann - can you please run it for some time *without* SIMPLE_FILES?

**Comment by CaBeckmann** [comment]
Sure, I can try it tomorrow.

**Comment by vanhauser-thc** [comment]
@CaBeckmann after this last issue is cleared up we can to make a new release ... :)

**Comment by CaBeckmann** [comment]
All instances are still running after 5 hours even without SIMPLE_FILES defined. If this is long enough for you I think we can consider this fixed and close this issue. If not I am letting this run until tomorrow anyway. Just let me know and I can post another update.

**Comment by vanhauser-thc** [comment]
thank you!

**Comment by hexcoder-** [comment]
@vanhauser-thc @andreafioraldi Just to understand, do you think, the filename corruption was because of the illegal combination of -L and -R?

**Comment by vanhauser-thc** [comment]
the filename corruption was because of a memory corruption in MOpt. then the tought was that there might be another issue with long filenames, but that was thankfully not the case.

**Comment by hexcoder-** [comment]
Thanks!
